Refresh.GameServer/Middlewares/DigestMiddleware.cs

```diff
@@ -0,0 +1,80 @@
+using System.Diagnostics;
+using System.Security.Cryptography;
+using Bunkum.CustomHttpListener.Request;
+using Bunkum.HttpServer;
+using Bunkum.HttpServer.Database;
+using Bunkum.HttpServer.Endpoints.Middlewares;
+using Refresh.GameServer.Extensions;
+
+namespace Refresh.GameServer.Middlewares;
+
+public class DigestMiddleware : IMiddleware
+{
+    // Should be 19 characters (or less maybe?)
+    // Length was taken from PS3 and PS4 digest keys
+    private const string DigestKey = "CustomServerDigest";
+
+    private string CalculateDigest(string url, Stream body, string auth)
+    {
+        using MemoryStream ms = new();
+
+        // FIXME: Directly referencing LBP in Bunkum
+        if (!url.StartsWith("/lbp/upload/"))
+        {
+            // get request body
+            body.CopyTo(ms);
+            body.Seek(0, SeekOrigin.Begin);
+        }
+
+        ms.WriteString(auth);
+        ms.WriteString(url);
+        ms.WriteString(DigestKey);
+
+        ms.Position = 0;
+        using SHA1 sha = SHA1.Create();
+        string digestResponse = Convert.ToHexString(sha.ComputeHash(ms)).ToLower();
+
+        return digestResponse;
+    }
+    
+    // Referenced from Project Lighthouse
+    // https://github.com/LBPUnion/ProjectLighthouse/blob/d16132f67f82555ef636c0dabab5aabf36f57556/ProjectLighthouse.Servers.GameServer/Middlewares/DigestMiddleware.cs
+    // https://github.com/LBPUnion/ProjectLighthouse/blob/19ea44e0e2ff5f2ebae8d9dfbaf0f979720bd7d9/ProjectLighthouse/Helpers/CryptoHelper.cs#L35
+    // TODO: make this non-lbp specific, or implement middlewares and move to game server
+    private bool VerifyDigestRequest(ListenerContext context)
+    {
+        string url = context.Uri.AbsolutePath;
+        string auth = $"{context.Cookies["MM_AUTH"] ?? string.Empty}";
+    
+        string digestResponse = this.CalculateDigest(url, context.InputStream, auth);
+    
+        string digestHeader = !url.StartsWith("/lbp/upload/") ? "X-Digest-A" : "X-Digest-B";
+        string clientDigest = context.RequestHeaders[digestHeader] ?? string.Empty;
+        
+        context.ResponseHeaders["X-Digest-B"] = digestResponse;
+        if (clientDigest == digestResponse) return true;
+        
+        // this._logger.LogWarning(BunkumContext.Digest, $"Digest failed: {clientDigest} != {digestResponse}");
```

Comment:
we should log these to help people with lbp3 (or was it lbp2 that complains with wrong digest key), perhaps we should even save failed digest requests and display them to the user on the website to say "hey you patched incorrectly!" but this is more a "feature request" than a problem

Reply:
We can probably bundle this with Recent Activity / Tokens (tracked @ https://github.com/LittleBigRefresh/Refresh/issues/16 and https://github.com/LittleBigRefresh/Refresh/issues/19 respectively) and make that system a sort of "event log" that a user can view.

But yeah this isn't relevant to the PR so I'll resolve this lol
Feel free to note this idea in one of those issues